# Previous

- CompletableFuture 는 비동기 작업의 대기, 취소를 위한 메서드를 제공하고 있다.

![image](https://github.com/shin-je-woo/TIL/assets/39439576/77685b40-55c0-47b7-a3d6-7aef6c2b1d78)

# 💡 get() / get(timeout, unit)

- 작업이 완료될 때 까지 혹은 지정된 시간까지 대기하고 결과를 반환한다.
- 예외로 작업이 완료된 경우 CancellationException(취소예외), ExecutionException(실행예외), InterruptedException(인터럽트예외) 를 발생시키며 예외를 처리해 주어야 한다.

```java
CompletableFuture<Integer> future = CompletableFuture.supplyAsync(() -> {
    throw new RuntimeException("작업 실패!");
});
try {
    Integer result = future.get(); // 체크드 예외로써 발생할 수 있는 모든 예외에 대한 처리가 필요하다
    System.out.println("결과: " + result);
} catch (InterruptedException e) {
    System.err.println("인트럽트 발생");
} catch (ExecutionException e) {
    System.err.println("실행 예외 발생");
}
```

# 💡 join()

- 작업이 완료될 때 까지 대기하고 결과를 반환한다.
- 예외로 작업이 완료된 경우 CancellationException(취소예외), CompletionException(완료예외) 발생시키며 언체크드 예외로서 예외 처리를 해 주지 않아도 된다.

```java
CompletableFuture<Integer> future = CompletableFuture.supplyAsync(() -> {
    throw new RuntimeException("작업 실패!");
});
Integer result = future.join(); // 언체크드 예외로써 발생하는 예외에 대한 처리는 필요 없지만 별도의 대응은 필요하다
System.out.println("결과: " + result);
```
